cpp/src/interop/from_arrow_device.cu

```diff
@@ -0,0 +1,448 @@
+  rmm::mr::device_memory_resource* mr)
+{
+  auto offsets_view = column_view{data_type(type_id::INT32),
+                                  static_cast<size_type>(input->offset + input->length + 1),
+                                  input->buffers[1],
+                                  nullptr,
+                                  0,
+                                  0};
+
+  ArrowSchemaView child_schema_view;
+  NANOARROW_THROW_NOT_OK(
+    ArrowSchemaViewInit(&child_schema_view, schema->schema->children[0], nullptr));
+  auto child_type = arrow_to_cudf_type(&child_schema_view);
+  auto [child_view, owned] =
+    get_column(&child_schema_view, input->children[0], child_type, false, stream, mr);
+
+  // in the scenario where we were sliced and there are more elements in the child_view
+  // than can be referenced by the sliced offsets, we need to slice the child_view
+  // so that when `get_sliced_child` is called, we still produce the right result
+  auto max_child_offset =
+    cudf::detail::get_value<int32_t>(offsets_view, input->offset + input->length, stream);
+  child_view = cudf::slice(child_view, {0, max_child_offset}, stream).front();
+
+  return std::make_tuple<column_view, owned_columns_t>(
+    {type,
+     static_cast<size_type>(input->length),
+     rmm::device_buffer{0, stream, mr}.data(),
+     reinterpret_cast<bitmask_type const*>(input->buffers[0]),
+     static_cast<size_type>(input->null_count),
+     static_cast<size_type>(input->offset),
+     {offsets_view, child_view}},
+    std::move(owned));
+}
+
+std::tuple<column_view, owned_columns_t> get_column(ArrowSchemaView* schema,
+                                                    const ArrowArray* input,
+                                                    data_type type,
+                                                    bool skip_mask,
+                                                    rmm::cuda_stream_view stream,
+                                                    rmm::mr::device_memory_resource* mr)
+{
+  return type.id() != type_id::EMPTY
+           ? std::move(type_dispatcher(
+               type, dispatch_to_cudf_column{}, schema, input, type, skip_mask, stream, mr))
+           : std::make_tuple<column_view, owned_columns_t>(get_empty_type_column(input->length),
+                                                           {});
+}
+
+}  // namespace
+
+unique_table_view_t from_arrow_device(ArrowSchemaView* schema,
+                                      const ArrowDeviceArray* input,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::mr::device_memory_resource* mr)
+{
+  if (input->sync_event != nullptr) {
+    CUDF_CUDA_TRY(
+      cudaStreamWaitEvent(stream.value(), *reinterpret_cast<cudaEvent_t*>(input->sync_event)));
+  }
+
+  std::vector<column_view> columns;
+  owned_columns_t owned_mem;
+
+  auto type = arrow_to_cudf_type(schema);
+  if (type != data_type(type_id::STRUCT)) {
```

Comment:
IIUC the reason that you have this block is so that a caller could provide either a top-level struct column representing multiple columns (i.e. a table) with each child representing a column in the table, or they could provide a single non-struct column in which case we interpret that as a single column to stick into a table and return. Is that what you're trying to accomplish here? Please let me know if I'm misunderstanding.

Assuming that I've understood correctly, I don't think that we should support this use case. This breaks some key invariants, such as the idempotency of `to_arrow_device(from_arrow_device(...))`, and it removes the otherwise fairly strict isomorphism between an ArrowDeviceArray and a cudf table_view (up to violations that come from `to_arrow_device` transforming 32 and 64 bit decimals to 128 bit). Moreover, it makes it practically confusing that a single struct column cannot be interconverted whereas a single column of any other type can.

I recognize the dilemma here though in that for `to_arrow_device` we could have column and table overloads whereas here we cannot return based solely on the return type. Although none of the alternatives are ideal here, I would generally prefer one of the obvious ones: 1) add a parameter, 2) add a template parameter, or 3) add a separate function with a different name (e.g. `from_arrow_device_column` and `from_arrow_device_table`).

Reply:
Of the options, I guess the best option I can think of would be to add a `from_arrow_device_column` function and have the `from_arrow_device` function that returns a table_view throw an error if the input is not a struct column.

Reply:
I think that would be fine. IMO that is preferable to the current behavior. @davidwendt WDYT?

Reply:
I agree with adding the `from_arrow_device_column` and throwing an exception if passed to the wrong API.

Reply:
Cool let's do that then.